src/db/gc/mod.rs

```diff
@@ -0,0 +1,323 @@
+    marked: HashSet<u32>,
+    epoch_marked: ChainEpoch,
+    depth: ChainEpochDelta,
+    block_time: Duration,
+}
+
+impl<DB: Blockstore + GarbageCollectable, GT: Fn() -> Arc<Tipset>> MarkAndSweep<DB, GT> {
+    /// Creates a new mark-and-sweep garbage collector.
+    ///
+    /// # Arguments
+    ///
+    /// * `db` - A reference to the database instance.
+    /// * `get_heaviest_tipset` - A function that facilitates heaviest tipset retrieval.
+    /// * `depth` - The number of state-roots to retain.
+    /// * `block_time` - An average block production time.
+    pub fn new(
+        db: Arc<DB>,
+        get_heaviest_tipset: GT,
+        depth: ChainEpochDelta,
+        block_time: Duration,
+    ) -> Self {
+        Self {
+            db,
+            get_heaviest_tipset,
+            depth,
+            marked: HashSet::new(),
+            epoch_marked: 0,
+            block_time,
+        }
+    }
+    // Populate the initial set with all the available database keys.
+    fn populate(&mut self) -> anyhow::Result<()> {
+        self.marked = self.db.get_keys()?;
+        Ok(())
+    }
+
+    // Filter out the initial set, leaving only the entries that need to be removed.
+    // NOTE: One concern here is that this is going to consume a lot of CPU.
+    async fn filter(&mut self, tipset: Arc<Tipset>, depth: ChainEpochDelta) -> anyhow::Result<()> {
+        // NOTE: We want to keep all the block headers from genesis to heaviest tipset epoch.
+        let mut stream = stream_graph(
+            self.db.clone(),
+            (*tipset).clone().chain(self.db.clone()),
+            depth,
+        );
+
+        while let Some(block) = stream.next().await {
+            let block = block?;
+            self.marked.remove(&truncated_hash(block.cid.hash()));
+        }
+
+        anyhow::Ok(())
+    }
+
+    // Remove marked keys from the database.
+    fn sweep(&mut self) -> anyhow::Result<()> {
+        let marked = mem::take(&mut self.marked);
+        self.db.remove_keys(marked)
+    }
+
+    /// Starts the Garbage Collection loop.
+    ///
+    /// # Arguments
+    ///
+    /// * `depth` - Specifies how far back the full history should be maintained. Cannot be less
+    /// than chain finality.
+    /// * `interval` - GC Interval to avoid constantly consuming node's resources.
+    ///
+    /// NOTE: This currently does not take into account the fact that we might be starting the node
+    /// using CAR-backed storage with a snapshot, for implementation simplicity.
+    pub async fn gc_loop(&mut self, interval: Duration) -> anyhow::Result<()> {
+        loop {
+            match self.gc_workflow().await? {
+                ControlFlow::Continue => continue,
+                ControlFlow::Finished => {
+                    // Make sure we don't run the GC too often.
+                    time::sleep(interval).await;
+                }
+            }
+        }
+    }
+
+    async fn gc_workflow(&mut self) -> anyhow::Result<ControlFlow> {
```

Comment:
I find the `ControlFlow` part a bit confusing. It appears you're only using it for testing. I don't understand why, though. Let's discuss this in Slack tomorrow.

Reply:
It's not for testing, it's used in order to be able to know whether or not we have to wait `interval` before restarting the cycle.

Reply:
Why not have `gc_workflow` always complete one iteration, then?

Reply:
> Why not have `gc_workflow` always complete one iteration, then?

It felt easier at the time to yield to the main loop. But with removal of `manual gc` this might not be
needed anymore.

Reply:
After removing `ControlFlow` this seems easier to read. I still stand by the current approach that yields to the main loop as it's much cleaner than having to write inner loops and introducing code repetition and mutable variables.